### PR TITLE
feat(conditions): Add JsonCondition for direct JSON evaluation

### PR DESCRIPTION
**Pull request by theref**
## Summary

This PR introduces **JsonCondition**, a new condition type that evaluates JSON data directly without requiring an HTTP request. This provides a simpler, more general alternative to `JsonApiCondition` for cases where JSON data is already available.

## Motivation

Currently, `JsonApiCondition` is designed for fetching JSON from HTTPS endpoints and then querying it. However, there are use cases where:
- JSON data is already available via Context
- Data comes from Sequential Condition

`JsonCondition` fills this gap by providing direct JSON evaluation with optional JSONPath queries.

### Core Implementation
1. **New `JsonCondition` class** (`nucypher/policy/conditions/json/json.py`)
   - Accepts JSON as dict, list, primitive, or JSON string
   - Optional JSONPath query support
   - Context variable resolution
   - Simpler architecture (inherits from `Condition` directly, not `ExecutionCallCondition`)
   - Uses standard `AnyField` for serialization

2. **Shared query logic** (`nucypher/policy/conditions/json/utils.py`)
   - Extracted `query_json_data()` utility function
   - Eliminates code duplication between `JsonCondition` and `JsonApiCondition`
   - Both condition types now use the same JSONPath query logic

3. **Type system integration**
   - Added `ConditionType.JSON` enum value
   - Added `JsonConditionDict` type definition
   - Registered in `ConditionLingo.resolve_condition_class()`

### Testing
- 11 comprehensive unit tests covering all functionality
- Tests include: initialization, validation, query execution, context variables, serialization, edge cases
- All existing tests still pass (351/351 condition tests passing)

## Usage Example

```python
from nucypher.policy.conditions.json.json import JsonCondition
from nucypher.policy.conditions.lingo import ReturnValueTest

# With dict and JSONPath query
condition = JsonCondition(
    data={"store": {"book": [{"price": 10.5}]}},
    query="$.store.book[0].price",
    return_value_test=ReturnValueTest("==", 10.5)
)

# With JSON string
condition = JsonCondition(
    data='{"value": 42}',
    query="$.value",
    return_value_test=ReturnValueTest("==", 42)
)

# With primitive (no query needed)
condition = JsonCondition(
    data=100,
    return_value_test=ReturnValueTest(">", 50)
)

# Verify
success, value = condition.verify()
```

### Atomic Commits
The changes are organized into 4 logical commits:
1. **Refactor**: Extract shared JSON query logic
2. **Feature**: Add JsonCondition implementation
3. **Chore**: Register in type system
4. **Test**: Add comprehensive tests

Each commit is self-contained and can be reviewed independently.

### Code Quality
- ✅ All pre-commit hooks passing (ruff, darker, etc.)
- ✅ No regressions - all 351 existing condition tests pass
- ✅ Full test coverage for new functionality
- ✅ Follows existing patterns and conventions

### Backward Compatibility
- ✅ No breaking changes to existing APIs
- ✅ `JsonApiCondition` behavior unchanged (just uses shared utility)
- ✅ All existing tests pass without modification

## Related Issues

This implements a more basic, general JSON condition type as discussed in the context of making TACo conditions more flexible and composable.

## Checklist

- [x] Code follows project style guidelines
- [x] Tests added and passing
- [x] No breaking changes
- [x] Documentation updated (newsfragment included)
- [x] Pre-commit hooks passing

### CODE REVIEW

File: nucypher/policy/conditions/json/json.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+def _parse_json_data(data: Any) -> Any:
+    """
+    Parse JSON data if it's a string, otherwise return as-is.
+    """
+    if not isinstance(data, str):
+        return data
+
+    try:
+        return json.loads(data)
+    except (json.JSONDecodeError, ValueError) as e:
```
Comment (derekpierre) [inline nucypher/policy/conditions/json/json.py]: `JSONDecodeError` is already a subclass of `ValueError`

```suggestion
    except (ValueError) as e:
```

File: nucypher/policy/conditions/json/json.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    An optional JSONPath query can be applied to extract a specific value.
+    """
+
+    CONDITION_TYPE = ConditionType.JSON.value
+
+    class Schema(Condition.Schema):
+        condition_type = fields.Str(
+            validate=validate.Equal(ConditionType.JSON.value), required=True
+        )
+        data = AnyField(required=True)
```
Comment (derekpierre) [inline nucypher/policy/conditions/json/json.py]: Would this simply always be a context variable? 🤔 

Analogous to `authorization_token` for `JsonApiCall` - https://github.com/nucypher/nucypher/blob/v7.7.x/nucypher/policy/conditions/json/api.py#L49

File: nucypher/policy/conditions/json/json.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        query = JSONPathField(required=False, allow_none=True)
```
Comment (derekpierre) [inline nucypher/policy/conditions/json/json.py:36]: Should this be required? Is there a case for it not needing to be used?

It wasn't required for JSON Rest endpoints because those can return values and not JSON.

If you deem it required, then update the constructor type to not be Optional.

File: nucypher/policy/conditions/json/json.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        result = query_json_data(resolved_data, self.query, **context)
+
+        # Process result for evaluation (handles string quoting)
+        result_for_eval = process_result_for_condition_eval(result)
+
+        # Evaluate against return value test
+        resolved_return_value_test = self.return_value_test.with_resolved_context(
+            **context
+        )
+        eval_result = resolved_return_value_test.eval(result_for_eval)
```
Comment (derekpierre) [inline nucypher/policy/conditions/json/json.py:90]: 🚀

File: nucypher/policy/conditions/types.py [hunk 5]
```diff
@@ -60,6 +60,11 @@ class ContractConditionDict(RPCConditionDict):
     functionAbi: NotRequired[ABIFunction]
 
 
+class JsonConditionDict(BaseExecConditionDict):
+    data: Any
+    query: NotRequired[str]
```
Comment (derekpierre) [inline nucypher/policy/conditions/types.py:65]: Update if `query` deemed required.

**Review by derekpierre — COMMENTED** [review]
Nice work - some questions/suggestions.

**Comment by derekpierre** [comment]
Let's see if we can cover most/all of the missing 6 lines for code coverage in json/utils.py 😉 :

Screenshot 2025-09-30 at 1 57 30 PM

(Assuming the coverage report is correct, I'm surprised those lines aren't covered by other tests).

File: nucypher/policy/conditions/json/json.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    process_result_for_condition_eval,
+    query_json_data,
+)
+from nucypher.policy.conditions.lingo import AnyField, ConditionType, ReturnValueTest
+
+
+def _parse_json_data(data: Any) -> Any:
+    """
+    Parse JSON data if it's a string, otherwise return as-is.
+    """
```
Comment (manumonti) [inline nucypher/policy/conditions/json/json.py]: I'm wondering if we could manage data as a string (not a JSON string). For example, data is an address `data = "0xc0ffee254729296a45a3885639AC7E10F9d54979"`, or `data = "enabled"`

**Review by manumonti — APPROVED** [review]
Looks great to me. I have a suggestion though. 

BTW, I'm super fan of how this PR looks. Commits are clear and concise, with well-formed messages. The PR message adds all the necessary context and makes clear the purpose of the changes and how they were made. Awesome.

File: nucypher/policy/conditions/json/json.py (hunk 2 above)
Comment (theref) [reply nucypher/policy/conditions/json/json.py]: you're right, I can't think of any situation where this wouldn't be populated by context in some way. For some reason I don't want to enforce that though

File: nucypher/policy/conditions/json/json.py (hunk 3 above)
Comment (theref) [reply nucypher/policy/conditions/json/json.py:36]: The JSON could also just return a value, I think there's a test for that...if there's not I'll add one

File: nucypher/policy/conditions/json/json.py (hunk 2 above)
Comment (derekpierre) [reply nucypher/policy/conditions/json/json.py]: > For some reason I don't want to enforce that though

Could you expand on your apprehension? If not a context variable, it means that it could be hardcoded - is there a legit case for that?

File: nucypher/policy/conditions/json/json.py (hunk 3 above)
Comment (derekpierre) [reply nucypher/policy/conditions/json/json.py:36]: Fair point. I guess in the case of it just forwarding along a JSON response it could just be value.

There's an prior test for JsonApiCondition for just a value - see https://github.com/nucypher/nucypher/blob/main/tests/unit/conditions/test_json_api_condition.py#L57

File: nucypher/policy/conditions/json/json.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    try:
+        return json.loads(data)
+    except ValueError as e:
+        raise InvalidCondition(f"Invalid JSON string: {e}") from e
+
+
+class JsonCondition(Condition):
+    """
+    A JSON condition evaluates data that is directly provided as JSON.
+    The data can be a dict, list, primitive value, or a JSON string.
```
Comment (cygnusv) [inline nucypher/policy/conditions/json/json.py]: I don't understand the motivation for this. Shouldn't we restrict this to strings?

File: tests/unit/conditions/test_json_condition.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+    assert condition.data == {"value": 42}
+
+
+def test_json_condition_with_primitive():
+    condition = JsonCondition(
+        data=42,
+        return_value_test=ReturnValueTest("==", 42),
+    )
+    assert condition.data == 42
+    assert condition.query is None
```
Comment (cygnusv) [inline tests/unit/conditions/test_json_condition.py]: Related to my comment above. I know that `"42"` is a valid JSON string, but `42` is just an integer.

Comment (theref) [reply tests/unit/conditions/test_json_condition.py]: I thnk `42` is also valid json, it's certainly accepted by a lot of json parsers, including the one we use.

File: nucypher/policy/conditions/json/json.py (hunk 6 above)
Comment (theref) [reply nucypher/policy/conditions/json/json.py]: those are both valid json, so would be handled fine

File: nucypher/policy/conditions/json/json.py (hunk 7 above)
Comment (theref) [reply nucypher/policy/conditions/json/json.py]: it's so that we can handle sequential conditions, where the json comes in already parsed as an object

Comment (cygnusv) [reply nucypher/policy/conditions/json/json.py]: Oh I see. Thanks for the explanation 👍

File: tests/unit/conditions/test_json_condition.py (hunk 8 above)
Comment (cygnusv) [reply tests/unit/conditions/test_json_condition.py]: Right, my point was more about acceptable types (str vs integers), but since you mentioned before that with sequential conditions JSON data may be already parsed as objects, then I have no further comment 👍

**Review by cygnusv — APPROVED** [review]

File: nucypher/policy/conditions/json/json.py (hunk 2 above)
Comment (theref) [reply nucypher/policy/conditions/json/json.py]: i think it's just overly restrictive - what do we gain by restricting it to only context variables?

Comment (cygnusv) [reply nucypher/policy/conditions/json/json.py]: As a reference, the `jwt_token` field in `JwtCondition` is required to be a context variable.

Comment (theref) [reply nucypher/policy/conditions/json/json.py]: ok, didn't know that :) I don't feel overly strongly on this, so I'm happy to defer and follow you both

Comment (derekpierre) [reply nucypher/policy/conditions/json/json.py]: > i think it's just overly restrictive - what do we gain by restricting it to only context variables?

One gain is avoiding backwards compatibility issues.

It's easier to go from allowing only context variable to allowing anything, vs allowing anything (and potential user misuse) and then trying to restrict to only context variables.

**Review by derekpierre — APPROVED** [review]
🎸
